### DISCUSSION TITLE
Standby, Sleep, Lowpower modes

### DISCUSSION THREAD

**Discussion by rjdza** [SpenceKonde/megaTinyCore#406] [Q&A]
Hi there

I'd like to be able to send the chip into low power mode for X seconds.

The research I did said to use lowpower.h, but compiling gives me a "LowPower.h: No such file or directory" error.

I can't find anything to help with this after a few searches foe sleep, low power, standby, ..., and I haven't had a chance to search thoroughly (it's for a mother's day project I'm making for my wife, and I only got the chips (ATtiny412) working today).

Any help appreciated.

Abstract: I've made a resin heart with an LED embedded inside it.  The LED is going to flash (Lub Dub) every 20 90 seconds.  It's designed as a pendant to be worn around the neck, and will run from a 3V CR2032.  The low power mode is required to help the battery to run as long as possible.

Everything is working except the low power standby / sleep. And trouble with the battery.

**Comment by pfoun** [comment]
@rjdza Here is a link to the document PowerSave.md which give solutions for power saving on the ATtiny.
https://github.com/SpenceKonde/megaTinyCore/blob/master/megaavr/extras/PowerSave.md

**Reply by rjdza** [reply]
Thanks for the link, it was useful.  Please see my response below.

**Comment by SpenceKonde** [comment]
LowPower.h is a third party library that you were supposed to install.. But no matter - it doesn't support the new parts, only the old "classic" AVR ones (which are, frankly, a lot less exciting, and worse for low power stuff)

https://github.com/SpenceKonde/megaTinyCore/blob/master/megaavr/extras/PowerSave.md; 

See also https://github.com/SpenceKonde/megaTinyCore/issues/158 - there have long been plans to make a library for the modernAVR parts (this, MegaCoreX and DxCore).and it is very frequently requested, but I am always dealing with more urgent or more interesting prioblems so I haven['t had time to work on it. I never really worked with power management. I only made one battery operated device in tjhe past few years, so there's a process of learning how to use that sort of feature myself. 

CR2032's are some really dismal batteries... check the datasheets on them for them like what the current was when they measured the capacity... they are meant to supply very low current intermittently over long periods of time. I think it was something like this that one person had which, when they turned on the power, it took so long to rise to operating voltages that they asked for options for a longer SUT setting on the fuses. And the way a lot of people use them in Arduino-circles is way the hell outside their specs!

**Reply by rjdza** [reply]
Thanks for the link and info.

Is there currently a way for me to say "Fire an interrupt in x ms" so that I can control the length of the delay?  Currently I'm using sleep_cpu() in a loop to get the approximate time I want.

As to the batteries - it's for a necklace that my wife will wear around her neck.  The battery needs to be small enough that it is comfortable and not too heavy.  I'm heavily committed to the 2032, but will consider alternatives if you can suggest something.

The load isn't that great - Ideally, the LED will pulse for about 2 seconds every 40 seconds or so. This is all the MCU (ATtiny412) will do.  I would have preferred to use a flickering LED, but can't find anyone who sells them in my country.

**Comment by grandaspanna** [comment]
Something like this should work. I web-searched and stole/adapted the code. On my attiny1614 board, power during sleep was down to about 1uA. The CR2032's are not designed for this, but are commonly used for wearable LED projects. I think with a bit of tweaking, you should end up with something she'll like. This code just blinks PA3 - adjust as required.

Play with the RTC prescaler and CMP values to adjust timeouts.

```
#include <avr/io.h> 
#include <avr/interrupt.h>
#include <avr/sleep.h>
#include <util/delay.h>

//RTC ISR //////////////////////////////////////////////////////
ISR(RTC_CNT_vect)
{
  RTC.INTFLAGS = RTC_OVF_bm;
}

int main(void)
{
  //Configure pins: ////////////////////////////////////////
  
  // Set all pins to low power mode:
  for (uint8_t i = 0; i < 8; i++) {
    *((uint8_t *)&PORTA + 0x10 + i) |= 1 << PORT_PULLUPEN_bp;
  }

  for (uint8_t i = 0; i < 8; i++) {
    *((uint8_t *)&PORTB + 0x10 + i) |= 1 << PORT_PULLUPEN_bp;
  } 

  // Enable PA3 for ouput
  PORTA.DIR = PIN3_bm;

  // Pause for debugging/measuring
  _delay_ms(1000);
  
  //RTC INIT: //////////////////////////////////////////////

  while (RTC.STATUS > 0) {} // Wait for all register to be synchronized

  RTC.PER = 1024*10; //10 secs between wakes.
  RTC.INTCTRL = 0 << RTC_CMP_bp
  | 1 << RTC_OVF_bp; //Overflow interrupt.
  
  RTC.CTRLA = RTC_PRESCALER_DIV1_gc //NO Prescaler
  | 1 << RTC_RTCEN_bp         //Enable RTC
  | 1 << RTC_RUNSTDBY_bp;     //Run in standby

  RTC.CLKSEL = RTC_CLKSEL_INT1K_gc; // 32KHz divided by 32, i.e run at 1.024kHz

    
  //Enable interrupts //////////////////////////////////////
  sei();
  
  //Main loop: /////////////////////////////////////////////
  while (1) {
    set_sleep_mode(SLEEP_MODE_STANDBY);
    sleep_enable();
    sleep_cpu(); //Go into sleep
    
    //On awakening
    
    PORTA.OUTTGL = PIN3_bm;
    _delay_ms(2000); //Stay still for power measurement with DMM.
    PORTA.OUTTGL = PIN3_bm;
    
  }
}
```
